libbpf-tools/mysqld_qslower.bpf.c

```diff
@@ -0,0 +1,77 @@
+// SPDX-License-Identifier: GPL-2.0
+/* Copyright (c) 2022 Hengqi Chen */
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include <bpf/usdt.bpf.h>
+
+#include "mysqld_qslower.h"
+
+#define MAX_ENTRIES	10240
+
+const volatile __u64 lat_ns = 0;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, __u32);
+	__type(value, struct start);
+} starts SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(__u32));
+	__uint(value_size, sizeof(__u32));
+} events SEC(".maps");
+
+static int handle_start(const char *query)
```

Comment:
not sure why you need separate handle_start/handle_done, everything is pretty straightforward to keep in BPF_USDT(query_{start,done}) below. But this is minor.